File: backend/src/services/territoryView.service.ts
import { Prisma } from "@prisma/client";
import { prisma } from "../lib/prisma";
import { NeverSoldQuery, TerritoryViewQuery } from "../validators/territoryView.validators";
import { firstDayOfPeriod, lastDayOfPeriod, monthsInPeriod, monthsWhereOr, YearMonth } from "./period.util";
import { creditedRevenueByHospital, hospitalIdsWithCreditedSales, hospitalIdsWithSalesBySalesperson } from "./kpi.service";

// Module P1 (Territory KPI Rules ข้อ 8): the salesperson's territories come from assignments
// effective IN THE VIEWED PERIOD (same window as ownerNamesFor/activeOwnerCount), membership is
// territory-grain, and every amount uses revenue(T)'s ข้อ-2 math through SalesLineCredit — the
// shared helpers from kpi.service.ts, never a parallel definition.
export async function getTerritoryView(salespersonId: string, query: TerritoryViewQuery) {
  const salesperson = await prisma.salesperson.findUnique({ where: { id: salespersonId } });
  if (!salesperson) return null;

  const months: YearMonth[] = monthsInPeriod(query);
  const periodWindow = { effectiveFrom: { lte: lastDayOfPeriod(query) }, OR: [{ effectiveTo: null }, { effectiveTo: { gte: firstDayOfPeriod(query) } }] };
  const assignments = await prisma.territoryAssignment.findMany({ where: { salespersonId, ...periodWindow }, include: { territory: true } });
  const territoryIds = assignments.map((row) => row.territoryId);

  const fallback = salesperson.excludedFromTerritoryTotals || !territoryIds.length;
  const baseWhere: Prisma.SalesLineWhereInput = fallback ? {} : { hospital: { territoryId: { in: territoryIds } } };
  const productTypeFilter: Prisma.SalesLineWhereInput = query.productTypeId ? { productTypeId: query.productTypeId } : {};
  const scopedWhere = { ...baseWhere, ...productTypeFilter };

  // Membership + amounts both follow the credit lens: a territory's hospitals count whoever
  // holds non-excluded credit there (or, in OWN_CREDIT_ONLY / fallback mode, this person only).
  const ownCreditsOnly = fallback || query.creditOnly;
  const [currentIdsRaw, everIdsRaw, currentRevenue] = await Promise.all([
    ownCreditsOnly
      ? hospitalIdsWithSalesBySalesperson(salespersonId, scopedWhere, months)
      : hospitalIdsWithCreditedSales(scopedWhere, months),
    ownCreditsOnly
      ? hospitalIdsWithSalesBySalesperson(salespersonId, scopedWhere)
      : hospitalIdsWithCreditedSales(scopedWhere),
    ownCreditsOnly
      ? creditedSelfRevenueByHospital(salespersonId, scopedWhere, months)
      : creditedRevenueByHospital(scopedWhere, months),
  ]);

  const hospitals = await prisma.hospital.findMany({
    where: { id: { in: [...new Set([...currentIdsRaw, ...everIdsRaw])] } },
    select: { id: true, displayName: true, province: true },
  });
  const byId = new Map(hospitals.map((hospital) => [hospital.id, hospital]));

  const soldHospitals = [...currentIdsRaw]
    .map((hospitalId) => {
      const hospital = byId.get(hospitalId)!;
      return { hospital: { id: hospital.id, displayName: hospital.displayName }, revenue: currentRevenue.get(hospitalId) ?? 0 };
    })
    .sort((a, b) => b.revenue - a.revenue);
  const soldBeforeButNotInPeriod = [...everIdsRaw]
    .filter((id) => !currentIdsRaw.has(id))
    .map((hospitalId) => {
      const hospital = byId.get(hospitalId)!;
      return { hospital: { id: hospital.id, displayName: hospital.displayName, province: hospital.province ?? null } };
    })
    .sort((a, b) => a.hospital.displayName.localeCompare(b.hospital.displayName, "th"));

  return {
    salesperson: { id: salesperson.id, displayName: salesperson.displayName },
    territories: assignments.map((row) => ({ id: row.territory.id, displayName: row.territory.name })),
    mode: fallback ? "NATIONWIDE_PRODUCT_TYPE_FALLBACK" : query.creditOnly ? "OWN_CREDIT_ONLY" : "TERRITORY_TOTAL",
    creditOnly: query.creditOnly,
    productTypeId: query.productTypeId ?? null,
    soldHospitals,
    soldBeforeButNotInPeriod,
  };
}

async function creditedSelfRevenueByHospital(salespersonId: string, baseWhere: Prisma.SalesLineWhereInput, months?: YearMonth[]): Promise<Map<string, number>> {
  const rows = await prisma.salesLineCredit.findMany({
    where: { salespersonId, salesLine: { ...baseWhere, ...(months ? { OR: monthsWhereOr(months) } : {}) } },
    select: { sharePercent: true, salesLine: { select: { hospitalId: true, total: true } } },
  });
  const totals = new Map<string, number>();
  for (const row of rows) totals.set(row.salesLine.hospitalId, (totals.get(row.salesLine.hospitalId) ?? 0) + Number(row.salesLine.total) * Number(row.sharePercent) / 100);
  return totals;
}

// Module P2 (Territory KPI Rules ข้อ 8 / plan.md Phase 16):
// "โรงพยาบาลรัฐทั่วไป (GOVERNMENT_GENERAL) ที่ยังไม่เคยมีรายการขายเลย" ในเขตของพนักงานขาย
// จำกัดจำนวนด้วย 2 กลไกพร้อมกันเสมอ: Top N ตามเกณฑ์ศักยภาพ และตัวกรองจังหวัด
export async function getNeverSoldHospitals(salespersonId: string, query: NeverSoldQuery) {
  const salesperson = await prisma.salesperson.findUnique({ where: { id: salespersonId } });
  if (!salesperson) return null;

  const periodWindow = { effectiveFrom: { lte: lastDayOfPeriod(query) }, OR: [{ effectiveTo: null }, { effectiveTo: { gte: firstDayOfPeriod(query) } }] };
  const assignments = await prisma.territoryAssignment.findMany({ where: { salespersonId, ...periodWindow }, include: { territory: true } });
  const territoryIds = assignments.map((row) => row.territoryId);

  const fallback = Boolean(salesperson.excludedFromTerritoryTotals || !territoryIds.length);

  // หา hospitalId ทั้งหมดที่มีประวัติการขาย (ถ้ามี productTypeId ให้กรองเฉพาะกลุ่มสินค้านั้น)
  const soldLines = await prisma.salesLine.findMany({
    where: query.productTypeId ? { productTypeId: query.productTypeId } : {},
    select: { hospitalId: true },
    distinct: ["hospitalId"],
  });
  const soldHospitalIds = soldLines.map((line) => line.hospitalId);

  // หา hospitalRegistryId ที่ผูกกับโรงพยาบาลที่เคยขาย
  const linkedSoldRegistries = await prisma.hospitalRegistryLink.findMany({
    where: { hospitalId: { in: soldHospitalIds }, status: "LINKED", hospitalRegistryId: { not: null } },
    select: { hospitalRegistryId: true },
  });
  const soldRegistryIds = new Set(linkedSoldRegistries.map((r) => r.hospitalRegistryId!));

  // ดึงโรงพยาบาลรัฐทั่วไปในทะเบียนที่ยังไม่เคยมีรายการขาย
  const registries = await prisma.hospitalRegistry.findMany({
    where: {
      category: "GOVERNMENT_GENERAL",
      isActive: true,
      id: { notIn: [...soldRegistryIds] },
      ...(fallback ? {} : { territoryId: { in: territoryIds } }),
      ...(query.provinceMappingId ? { provinceMappingId: query.provinceMappingId } : {}),
    },
    include: {
      provinceMapping: { select: { id: true, canonicalName: true } },
      territory: { select: { id: true, name: true } },
      metrics: {
        where: { metric: query.potentialMetric },
        select: { value: true },
        orderBy: { createdAt: "desc" },
        take: 1,
      },
    },
  });

  const formatted = registries.map((reg) => {
    const metricVal = reg.metrics.length ? Number(reg.metrics[0].value) : 0;
    return {
      id: reg.id,
      displayName: reg.displayName,
      province: reg.provinceMapping?.canonicalName ?? reg.provinceRaw,
      provinceMappingId: reg.provinceMappingId ?? null,
      tier: reg.tier ?? null,
      category: reg.category,
      metricKey: query.potentialMetric,
      metricValue: metricVal,
      territory: reg.territory ? { id: reg.territory.id, displayName: reg.territory.name } : null,
    };
  });


  formatted.sort((a, b) => {
    if (b.metricValue !== a.metricValue) return b.metricValue - a.metricValue;
    return a.displayName.localeCompare(b.displayName, "th");
  });

  const sliced = formatted.slice(0, query.topN);

  return {
    salesperson: { id: salesperson.id, displayName: salesperson.displayName },
    territories: assignments.map((row) => ({ id: row.territory.id, displayName: row.territory.name })),
    mode: fallback ? "NATIONWIDE_PRODUCT_TYPE_FALLBACK" : "TERRITORY_TOTAL",
    potentialMetric: query.potentialMetric,
    topN: query.topN,
    provinceMappingId: query.provinceMappingId ?? null,
    productTypeId: query.productTypeId ?? null,
    totalNeverSold: formatted.length,
    neverSoldHospitals: sliced,
  };
}

File: src/SalesEvaluation.Application/Territories/TerritoryViewService.cs
namespace SalesEvaluation.Application.Territories;

using System.Globalization;
using Microsoft.EntityFrameworkCore;
using SalesEvaluation.Application.Common.Interfaces;
using SalesEvaluation.Contracts.TerritoryViews;
using SalesEvaluation.Domain.Entities;
using SalesEvaluation.Domain.Enums;

public class TerritoryViewService : ITerritoryViewService
{
    private static readonly StringComparer ThaiNameComparer = StringComparer.Create(new CultureInfo("th-TH"), ignoreCase: false);

    private readonly IAppDbContext _dbContext;

    public TerritoryViewService(IAppDbContext dbContext)
    {
        _dbContext = dbContext;
    }

    // ---- Shared period helpers (mirrors services/period.util.ts) ----

    internal static List<(int Year, int Month)> MonthsInPeriod(TerritoryViewPeriod period)
    {
        if (period.PeriodType == PeriodType.QUARTER)
        {
            var startMonth = (period.PeriodNumber - 1) * 3 + 1;
            return new List<(int, int)> { (period.Year, startMonth), (period.Year, startMonth + 1), (period.Year, startMonth + 2) };
        }

        if (period.PeriodType == PeriodType.YEAR)
        {
            return Enumerable.Range(1, 12).Select(m => (period.Year, m)).ToList();
        }

        return new List<(int, int)> { (period.Year, period.PeriodNumber) };
    }

    private static DateOnly FirstDayOfPeriod(TerritoryViewPeriod period)
    {
        var first = MonthsInPeriod(period)[0];
        return new DateOnly(first.Year, first.Month, 1);
    }

    private static DateOnly LastDayOfPeriod(TerritoryViewPeriod period)
    {
        var months = MonthsInPeriod(period);
        var last = months[^1];
        return new DateOnly(last.Year, last.Month, DateTime.DaysInMonth(last.Year, last.Month));
    }

    /// <summary>SalesLine (year, month) filter as a translatable month key.</summary>
    private static int MonthKey(int year, int month) => year * 12 + (month - 1);

    // ---- Module P1: territory view ----

    public async Task<MyTerritoryViewResponse?> GetTerritoryViewAsync(string salespersonId, TerritoryViewQuery query, CancellationToken cancellationToken = default)
    {
        var salesperson = await _dbContext.Salespeople
            .AsNoTracking()
            .FirstOrDefaultAsync(sp => sp.Id == salespersonId, cancellationToken);

        if (salesperson == null)
        {
            return null;
        }

        var months = MonthsInPeriod(query.Period);
        var monthKeys = months.Select(m => MonthKey(m.Year, m.Month)).ToList();

        // Module P1 (Territory KPI Rules ข้อ 8): territories come from assignments effective in
        // the viewed period (midnight-to-midnight comparison, so the boundary days are inclusive).
        var firstDay = FirstDayOfPeriod(query.Period);
        var lastDay = LastDayOfPeriod(query.Period);

        var assignments = await _dbContext.TerritoryAssignments
            .AsNoTracking()
            .Include(a => a.Territory)
            .Where(a => a.SalespersonId == salespersonId && a.EffectiveFrom <= lastDay && (a.EffectiveTo == null || a.EffectiveTo >= firstDay))
            .ToListAsync(cancellationToken);

        var territoryIds = assignments.Select(a => a.TerritoryId).ToList();

        var fallback = salesperson.ExcludedFromTerritoryTotals || territoryIds.Count == 0;
        var ownCreditsOnly = fallback || query.CreditOnly;

        var currentIds = await HospitalIdsWithCreditedSalesAsync(ownCreditsOnly, salespersonId, territoryIds, fallback, query.ProductTypeId, monthKeys, cancellationToken);
        var everIds = await HospitalIdsWithCreditedSalesAsync(ownCreditsOnly, salespersonId, territoryIds, fallback, query.ProductTypeId, monthKeys: null, cancellationToken);
        var currentRevenue = await CreditedRevenueByHospitalAsync(ownCreditsOnly, salespersonId, territoryIds, fallback, query.ProductTypeId, monthKeys, cancellationToken);

        var unionIds = currentIds.Union(everIds).ToList();
        var hospitals = await _dbContext.Hospitals
            .AsNoTracking()
            .Where(h => unionIds.Contains(h.Id))
            .Select(h => new { h.Id, h.DisplayName, h.Province })
            .ToDictionaryAsync(h => h.Id, cancellationToken);

        var soldHospitals = currentIds
            .Where(id => hospitals.ContainsKey(id))
            .Select(id => new SoldHospitalDto
            {
                Hospital = new EntitySummaryDto { Id = id, DisplayName = hospitals[id].DisplayName },
                Revenue = currentRevenue.GetValueOrDefault(id)
            })
            .OrderByDescending(row => row.Revenue)
            .ToList();

        var soldBeforeButNotInPeriod = everIds
            .Except(currentIds)
            .Where(id => hospitals.ContainsKey(id))
            .Select(id => new SoldBeforeHospitalDto
            {
                Hospital = new SoldBeforeHospitalNameDto
                {
                    Id = id,
                    DisplayName = hospitals[id].DisplayName,
                    Province = hospitals[id].Province
                }
            })
            .OrderBy(row => row.Hospital.DisplayName, ThaiNameComparer)
            .ToList();

        return new MyTerritoryViewResponse
        {
            Salesperson = new EntitySummaryDto { Id = salesperson.Id, DisplayName = salesperson.DisplayName },
            Territories = assignments.Select(a => new EntitySummaryDto { Id = a.Territory.Id, DisplayName = a.Territory.Name }).ToList(),
            Mode = fallback
                ? "NATIONWIDE_PRODUCT_TYPE_FALLBACK"
                : query.CreditOnly
                    ? "OWN_CREDIT_ONLY"
                    : "TERRITORY_TOTAL",
            CreditOnly = query.CreditOnly,
            ProductTypeId = query.ProductTypeId,
            SoldHospitals = soldHospitals,
            SoldBeforeButNotInPeriod = soldBeforeButNotInPeriod
        };
    }

    // ---- Module P2: never-sold government hospitals ----

    public async Task<NeverSoldHospitalsResponse?> GetNeverSoldHospitalsAsync(string salespersonId, NeverSoldQuery query, CancellationToken cancellationToken = default)
    {
        var salesperson = await _dbContext.Salespeople
            .AsNoTracking()
            .FirstOrDefaultAsync(sp => sp.Id == salespersonId, cancellationToken);

        if (salesperson == null)
        {
            return null;
        }

        var firstDay = FirstDayOfPeriod(query.Period);
        var lastDay = LastDayOfPeriod(query.Period);

        var assignments = await _dbContext.TerritoryAssignments
            .AsNoTracking()
            .Include(a => a.Territory)
            .Where(a => a.SalespersonId == salespersonId && a.EffectiveFrom <= lastDay && (a.EffectiveTo == null || a.EffectiveTo >= firstDay))
            .ToListAsync(cancellationToken);

        var territoryIds = assignments.Select(a => a.TerritoryId).ToList();
        var fallback = salesperson.ExcludedFromTerritoryTotals || territoryIds.Count == 0;

        // hospitalId ทั้งหมดที่มีประวัติการขาย (ถ้ามี productTypeId ให้กรองเฉพาะกลุ่มสินค้านั้น)
        var soldLinesQuery = _dbContext.SalesLines.AsNoTracking();
        if (!string.IsNullOrEmpty(query.ProductTypeId))
        {
            soldLinesQuery = soldLinesQuery.Where(sl => sl.ProductTypeId == query.ProductTypeId);
        }

        var soldHospitalIds = await soldLinesQuery
            .Select(sl => sl.HospitalId)
            .Distinct()
            .ToListAsync(cancellationToken);

        // hospitalRegistryId ที่ผูกกับโรงพยาบาลที่เคยขาย
        var soldRegistryIds = (await _dbContext.HospitalRegistryLinks
                .AsNoTracking()
                .Where(l => soldHospitalIds.Contains(l.HospitalId) && l.Status == RegistryLinkStatus.LINKED && l.HospitalRegistryId != null)
                .Select(l => l.HospitalRegistryId!)
                .ToListAsync(cancellationToken))
            .ToHashSet();

        // โรงพยาบาลรัฐทั่วไปในทะเบียนที่ยังไม่เคยมีรายการขาย
        var registriesQuery = _dbContext.HospitalRegistries
            .AsNoTracking()
            .Include(r => r.ProvinceMapping)
            .Include(r => r.Territory)
            .Include(r => r.Metrics.Where(m => m.Metric == query.PotentialMetric).OrderByDescending(m => m.CreatedAt).Take(1))
            .Where(r => r.Category == HospitalCategory.GOVERNMENT_GENERAL &&
                        r.IsActive &&
                        !soldRegistryIds.Contains(r.Id));

        if (!fallback)
        {
            registriesQuery = registriesQuery.Where(r => r.TerritoryId != null && territoryIds.Contains(r.TerritoryId));
        }

        if (!string.IsNullOrEmpty(query.ProvinceMappingId))
        {
            registriesQuery = registriesQuery.Where(r => r.ProvinceMappingId == query.ProvinceMappingId);
        }

        var registries = await registriesQuery.ToListAsync(cancellationToken);

        var formatted = registries.Select(r =>
        {
            var metricValue = r.Metrics.FirstOrDefault()?.Value ?? 0m;
            return new NeverSoldHospitalItemDto
            {
                Id = r.Id,
                DisplayName = r.DisplayName,
                Province = r.ProvinceMapping?.CanonicalName ?? r.ProvinceRaw,
                ProvinceMappingId = r.ProvinceMappingId,
                Tier = r.Tier,
                Category = r.Category.ToString(),
                MetricKey = query.PotentialMetric.ToString(),
                MetricValue = metricValue,
                Territory = r.Territory == null ? null : new EntitySummaryDto { Id = r.Territory.Id, DisplayName = r.Territory.Name }
            };
        })
        .OrderByDescending(item => item.MetricValue)
        .ThenBy(item => item.DisplayName, ThaiNameComparer)
        .ToList();

        return new NeverSoldHospitalsResponse
        {
            Salesperson = new EntitySummaryDto { Id = salesperson.Id, DisplayName = salesperson.DisplayName },
            Territories = assignments.Select(a => new EntitySummaryDto { Id = a.Territory.Id, DisplayName = a.Territory.Name }).ToList(),
            Mode = fallback ? "NATIONWIDE_PRODUCT_TYPE_FALLBACK" : "TERRITORY_TOTAL",
            PotentialMetric = query.PotentialMetric.ToString(),
            TopN = query.TopN,
            ProvinceMappingId = query.ProvinceMappingId,
            ProductTypeId = query.ProductTypeId,
            TotalNeverSold = formatted.Count,
            NeverSoldHospitals = formatted.Take(query.TopN).ToList()
        };
    }

    // ---- Shared credited-sales helpers (mirrors kpi.service.ts's single-source definitions) ----

    /// <summary>
    /// "Has credited sales" is contract-defined through SalesLineCredit so excluded personnel never
    /// count as territory sales. In own-credit mode only this person's shares count.
    /// </summary>
    private async Task<HashSet<string>> HospitalIdsWithCreditedSalesAsync(bool ownCreditsOnly, string salespersonId, List<string> territoryIds, bool fallback, string? productTypeId, List<int>? monthKeys, CancellationToken cancellationToken)
    {
        var query = CreditedSalesQuery(ownCreditsOnly, salespersonId, territoryIds, fallback, productTypeId, monthKeys);
        var ids = await query.Select(c => c.SalesLine.HospitalId).Distinct().ToListAsync(cancellationToken);
        return new HashSet<string>(ids);
    }

    private async Task<Dictionary<string, decimal>> CreditedRevenueByHospitalAsync(bool ownCreditsOnly, string salespersonId, List<string> territoryIds, bool fallback, string? productTypeId, List<int>? monthKeys, CancellationToken cancellationToken)
    {
        var rows = await CreditedSalesQuery(ownCreditsOnly, salespersonId, territoryIds, fallback, productTypeId, monthKeys)
            .Select(c => new { HospitalId = c.SalesLine.HospitalId, Total = c.SalesLine.Total, c.SharePercent })
            .ToListAsync(cancellationToken);

        var totals = new Dictionary<string, decimal>();
        foreach (var row in rows)
        {
            totals[row.HospitalId] = totals.GetValueOrDefault(row.HospitalId) + row.Total * (row.SharePercent / 100m);
        }

        return totals;
    }

    private IQueryable<SalesLineCredit> CreditedSalesQuery(bool ownCreditsOnly, string salespersonId, List<string> territoryIds, bool fallback, string? productTypeId, List<int>? monthKeys)
    {
        var query = _dbContext.SalesLineCredits.AsNoTracking().AsQueryable();

        if (ownCreditsOnly)
        {
            query = query.Where(c => c.SalespersonId == salespersonId);
        }
        else
        {
            query = query.Where(c => !c.Salesperson.ExcludedFromTerritoryTotals);
        }

        if (!fallback)
        {
            query = query.Where(c => c.SalesLine.Hospital != null && territoryIds.Contains(c.SalesLine.Hospital.TerritoryId!));
        }

        if (!string.IsNullOrEmpty(productTypeId))
        {
            query = query.Where(c => c.SalesLine.ProductTypeId == productTypeId);
        }

        if (monthKeys != null)
        {
            query = query.Where(c => monthKeys.Contains(c.SalesLine.Year * 12 + c.SalesLine.Month - 1));
        }

        return query;
    }
}
